docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,262 @@
+
+Currently, if LoRaDevice is not in the cache we search on the Function for devices with the specific
+DevAddr then we load their twins.
+
+We could ❔
+
+- Amend DeviceGetter.GetDevice so that it returns info about whether we are the winning gateway
+  - if we are (we already have the connection) continue to load the twin
+  - if we are not the winning LNS, drop the message and mark our selves as the losing gateway
+  - This would require moving the Mic computation on the Function (?)
+
+### Handling of ABP relax frame counter reset
+
+What happens in the above scenario, if we have a device reset between message A and B? Currently we
+are saving the twin immediately and then clear the Function cache.
+
+This is both dangerous as the request is not yet validated but is changing state of the twin as well
+as not required as at the end of processing the message we do sync changes on the twin.
+
+We should instead:
+
+- Clear the cache in the Function as we do currently.
+- Function returns if we are the losing LNS. If we are the losing one, we drop the message here,
+  mark ourselves as the losing gateway etc.
+- Otherwise we process message normally.
+- And at the end force update the twin.
+
+Alternative considered but discarded: use the Function to do both operations
+
+- update the device twin frame counter to 0 (it needs a DeviceClient that would
+  cause a temporary connection switch)
+- Clear or update the cache entry with frame counter down and up to 0.
+- Returns the result to the LNS: whether it was the winning or losing one
+- LNS reacts as described in the [main data flow section](#main-data-message-flow)
+
+### Main data message flow
+
+Assuming the topology:
+
+```mermaid
+flowchart LR;
+    Device-->LBS1-->LNS1-->Function;
+    Device-->LBS2-->LNS2-->Function;
+    LNS1-->IoTHub;
+    LNS2-->IoTHub;
+```
+
+where Device sends data message A and then B.
+
+Here is a rundown of what should happen marked in **bold**:
+
+1. Device sends first data message A.
+1. We assume that LNS1 gets the message first. **LNS1 checks against
+   an in-memory dictionary DevEui -> flag** and since it has not seen this DevEui before (flag is
+   false) contacts the GetDevice.
+1. The Function hasn't seen this DevEui either and therefore does not have an assigned LNS for it
+  yet. LNS1 wins the race and gets immediately a response and processes the message upstream.
+1. LNS2 eventually receives message A, **checks its local dictionary** and also contacts the FunctionBundler since it does not have prior info about this DevEui.
+1. The Function responds to LNS2 that it lost the race to process this message.
+1. Since deduplication strategy is Drop, LNS2 drops the message immediately, therefore no
+   connection to Iot Hub is opened and only LNS1 has the connection to Iot Hub. **LNS2 also notes in
+   memory that it was the losing gateway for this DevEui**.
+1. When message B gets send (with a higher frame counter), assuming that this time LNS2 gets it
```

Comment:
**I really love this scenario breakdown**

I think here we only consider the case with higher frame counter. However it could be cases where the frame counter could be lower (ABP Relax mode, or simply FCNT 32 bits rollover). I would suggest to also state in this flow when you actually check the FrameCounter. We should probably have a smart way of not dropping for wrong fcnt reason when you are the passive gateway.

Reply:
Yes, this flow assumes higher frame counter for later messages, the case of ABP relax mode is discussed in the previous section.

Do you think we should merge the 2 in one? My thinking of separating the 2 was to keep the main (or most common?) flow clean but we can change that. Or the feedback is something else?

Reply:
Edit: I think we are missing the resubmit case (frame counter equal to before) - I commented on another thread https://github.com/Azure/iotedge-lorawan-starterkit/pull/1483/files#r807395835

Reply:
Bug https://github.com/Azure/iotedge-lorawan-starterkit/issues/1468 is tracking the issue with resubmit

Reply:
Please let me know if there are other cases here missing.